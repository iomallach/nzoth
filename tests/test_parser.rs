mod helpers;
use std::cell::RefCell;

use helpers::SnapshotErrors;
use insta;
use nzoth::{parser::Parser, source::SourceFile};

#[test]
fn test_valid_let_declarations() {
    insta::glob!("resources/parser_cases/valid_let_declarations.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        let program = parser.parse();
        assert_eq!(2, program.nodes.len());

        let output = format!("{:#?}", program);

        insta::assert_snapshot!(output);
    })
}

#[test]
fn test_valid_identifier() {
    insta::glob!("resources/parser_cases/valid_identifiers.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        let program = parser.parse();

        let output = format!("{:#?}", program);
        insta::assert_snapshot!(output);
    })
}

#[test]
fn test_valid_assignments() {
    insta::glob!("resources/parser_cases/valid_assignments.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        let program = parser.parse();

        let output = format!("{:#?}", program);
        insta::assert_snapshot!(output);
    })
}

#[test]
fn test_invalid_assignments() {
    insta::glob!("resources/parser_cases/invalid_assignments.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        _ = parser.parse();

        insta::assert_snapshot!(SnapshotErrors(parser.errors()));
    })
}
