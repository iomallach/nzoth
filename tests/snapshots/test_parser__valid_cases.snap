---
source: tests/test_parser.rs
expression: output
input_file: tests/resources/parser_cases/valid_assignments.nz
---
Program {
    nodes: [
        Statement(
            Expression(
                ExpressionStatement {
                    span: Span {
                        start: Position {
                            offset: 0,
                            line: 0,
                            column: 0,
                        },
                        end: Position {
                            offset: 3,
                            line: 0,
                            column: 3,
                        },
                        file: FileId(
                            0,
                        ),
                    },
                    expression: Infix(
                        Identifier(
                            "foo",
                            Span {
                                start: Position {
                                    offset: 0,
                                    line: 0,
                                    column: 0,
                                },
                                end: Position {
                                    offset: 3,
                                    line: 0,
                                    column: 3,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        ),
                        Assignment,
                        Identifier(
                            "bar",
                            Span {
                                start: Position {
                                    offset: 6,
                                    line: 0,
                                    column: 6,
                                },
                                end: Position {
                                    offset: 9,
                                    line: 0,
                                    column: 9,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        ),
                        Span {
                            start: Position {
                                offset: 0,
                                line: 0,
                                column: 0,
                            },
                            end: Position {
                                offset: 3,
                                line: 0,
                                column: 3,
                            },
                            file: FileId(
                                0,
                            ),
                        },
                    ),
                },
            ),
        ),
        Statement(
            Expression(
                ExpressionStatement {
                    span: Span {
                        start: Position {
                            offset: 11,
                            line: 1,
                            column: 0,
                        },
                        end: Position {
                            offset: 14,
                            line: 1,
                            column: 3,
                        },
                        file: FileId(
                            0,
                        ),
                    },
                    expression: Infix(
                        Identifier(
                            "foo",
                            Span {
                                start: Position {
                                    offset: 11,
                                    line: 1,
                                    column: 0,
                                },
                                end: Position {
                                    offset: 14,
                                    line: 1,
                                    column: 3,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        ),
                        Assignment,
                        NumericLiteral(
                            Integer(
                                3,
                            ),
                            Span {
                                start: Position {
                                    offset: 17,
                                    line: 1,
                                    column: 6,
                                },
                                end: Position {
                                    offset: 18,
                                    line: 1,
                                    column: 7,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                        ),
                        Span {
                            start: Position {
                                offset: 11,
                                line: 1,
                                column: 0,
                            },
                            end: Position {
                                offset: 14,
                                line: 1,
                                column: 3,
                            },
                            file: FileId(
                                0,
                            ),
                        },
                    ),
                },
            ),
        ),
    ],
}
